birdhouse/config/thredds/default.env

```diff
@@ -6,10 +6,28 @@ export THREDDS_DOCKER=pavics/thredds-docker
 export THREDDS_IMAGE="${THREDDS_DOCKER}:${THREDDS_VERSION}"
 export THREDDS_ORGANIZATION="Birdhouse"
 
+# All default values for data locations are set to ensure backward compatibility
+export THREDDS_DATA_LOCATION_ON_CONTAINER='/pavics-ncml'
+export THREDDS_RAW_DATA_LOCATION_ON_CONTAINER='/pavics-data'
+export THREDDS_DATA_LOCATION_ON_HOST='${DATA_PERSIST_ROOT}/ncml'  # this default is for backward compatibility 
+export THREDDS_RAW_DATA_LOCATION_ON_HOST='${DATA_PERSIST_ROOT}/datasets'  # this default is for backward compatibility
+export THREDDS_DATA_LOCATION_NAME='Datasets'  # this default is for backward compatibility 
+export THREDDS_RAW_DATA_LOCATION_NAME='Birdhouse'  # this default is for backward compatibility 
+export THREDDS_DATA_LOCATION_NAME_lower='$(echo "$THREDDS_DATA_LOCATION_NAME" | tr "[:upper:]" "[:lower:]")'
+export THREDDS_RAW_DATA_LOCATION_NAME_lower='$(echo "$THREDDS_RAW_DATA_LOCATION_NAME" | tr '[:upper:]' '[:lower:]')'
```

Comment:
From the point of view of a new node manager, it is not clear what consists of a "`RAW_DATA"` vs plain `"DATA"`. Naming could be improved to provide better guidance.

From my understanding, the original 2 locations refer to something like:
- `birdhouse`: data shared amongst services of the server, such as user-workspaces data, test data, dynamically generated outputs from processing, etc.
- `datasets`: external data sources provided by the server, usually as read-only and static, provided for reference/research

Reply:
> From the point of view of a new node manager, it is not clear what consists of a "`RAW_DATA"` vs plain `"DATA"`. Naming could be improved to provide better guidance.

This was already discussed earlier in the PR and this naming convention was seen as the best path forward. I'm hesitant to re-visit it at this point. Regardless of the choice of name, we will have to document this properly for a new node admin, so that will be an appropriate venue to make the purpose of the two locations clear.

Reply:
I think the current names only swapped non-descriptive names (birdhouse, datasets) by new ones. 
These names do not improve item 2. of the PR description. 
If the admin has to find the comment that describes the variables (which are not present at the moment), it is already an indication that they are improperly named.

Reply:
> If the admin has to find the comment that describes the variables (which are not present at the moment), it is already an indication that they are improperly named.

I disagree. It is impossible to fully describe the purpose of the variable in a name. This is why we use documentation to properly describe the variable's purpose. 

> From the point of view of a new node manager, it is not clear what consists of a "RAW_DATA" vs plain "DATA". Naming could be improved to provide better guidance.

Ok, what would you suggest @fmigneault?

Reply:
The new names reflect the way in which data is currently stored on the thredds server on PAVICS, specifically, one folder for catalog data (now defined by the new variable `THREDDS_DATA_LOCATION_NAME`) and one for data that is not meant to be part of the official catalog  (now indicated by the variable name `THREDDS_RAW_DATA_LOCATION_NAME`). It seems very sensible to me given the dual needs of supporting legacy decisions while bringing much needed clarity and flexibility.

Reply:
IMO `datasets` and `birdhouse` made sense, because one is for storing static/external "datasets", while the other is for operational needs of the services under "birdhouse"-deploy. Maybe there was some misuses in the past in path names such as using `/pavics-data` or regarding were some data was placed for convenience instead of where they should actually be, but the original intention of these names were right.

Reply:
I agree that they make sense for developers who are organizing the code but these names are visible to users who are browsing the thredds catalog. A user doesn't know what a "birdhouse" is or how that relates to their data. 

I think we're getting off track here though since the default values "birdhouse" and "datasets" aren't changed by this PR, this PR just provides the option to set them to something different if a node administrator wants.

Reply:
> I agree that they make sense for developers who are organizing the code but these names are visible to users who are browsing the thredds catalog. A user doesn't know what a "birdhouse" is or how that relates to their data.

Yes, this is why these new variables allow to override the *values* to whichever name that feels right for your instance, such that what users will see is relevant on that server. The *name* of the variables themselves though should be meaningful for the devs, not the users that will not see those variable names anyway. It is irrelevant that a user knows about eg `THREDDS_BIRDHOUSE_DATA_LOCATION_NAME`/`THREDDS_DATASET_DATA_LOCATION_NAME` or not. The code base is "birdhouse-deploy", so I don't see what would be more relevant than "birdhouse" for a dev to indicate data directories used by its services. This would make it the same default name as the docker default network and prefix used in  named volumes automatically created by docker-compose.

Of course, the defaults *values* need to remain "birdhouse" and "datasets" respectively for backward compatibility. I'm only takling about the *name* of the new variables.

Reply:
The problem is the "Birdhouse" location, for historical reasons/misuse, is currently used for storing data used by services under birdhouse-deploy as well as for storing and exposing raw data that are not part of (the currently used) catalog, i.e. THREDDS. 

Ideally, there should be three directories: One strictly for storing catalog data, one strictly for data that for whatever reason one doesn't want to store in the catalog at the moment, and one for internal use by birdhouse.

Reply:
I am not opposed to this three directory structure. In fact, I'd prefer it. Then, we can have variables like: `THREDDS_BIRDHOUSE_DATA_LOCATION_NAME`, `THREDDS_RAW_DATA_LOCATION_NAME` and `THREDDS_DATA_LOCATION_NAME`.

Reply:
I'm not sure to undertand the use case of "data that for whatever reason one doesn't want to store in the catalog". 
If that data is not a dataset to be stored in the catalog, nor used by a service, what is it used for? I don't think the birdhouse-deploy config should have an unused variable. A service could however define more variables for additional data locations, such as what is done for `jupyterhub_user_data`, `magpie_persist`, etc., but we would add them as needed later on.

For the names, I would prefer `THREDDS_DATASET_LOCATION_NAME` or `THREDDS_CATALOG_LOCATION_NAME` (to make a parallel to your mention "storing *catalog* data"), but not simply `RAW_DATA` or `DATA`. These are too generic and not meaningful to disambiguate them, which will lead to more data locations being mixed/misused in the future.

Reply:
From my comment https://github.com/bird-house/birdhouse-deploy/pull/359#pullrequestreview-1540391553 explaining the historical reason for the naming of the 2 collections "Birdhouse" and "Datasets", those 2 names are, let's say, not the greatest choice.

So I think I'd rather not have "birdhouse" again in the variable name `THREDDS_BIRDHOUSE_DATA_LOCATION_NAME`.

The current "Birdhouse" catalog on our Thredds contains:
* `.nc` files that will be aggregated by an `.ncml` file in the "Datasets" catalog
* `.nc` files that we simply host for other orgs or other projects, ex: eccc/forecasts/geps/, that might or might not have a corresponding nicer `.ncml` aggregation in the "Datasets" catalog
* wps_outputs
* testdata

So it's pretty much a dumping ground for anything.  That "Birdhouse" catalog could be called "Root", "Misc" or "Raw" and it would still fit.

When we discovered all the features `.ncml` offers, we decided to create a brand new "Datasets" catalog to make it really stand out.

So from my perspective, `RAW_DATA` and `DATA` flavors for those 2 starting catalogs are fine for now.  Just document they are just 2 catalogs, nothing special, the admin will decide what goes in there.  It could be just testdata, just wps_outputs, not a mixed bag of everything like "Birdhouse" is right now.

What would really help future node admin is the ability to add new catalogs as they need, so they are not stuck with 2 like us and one of the 2 end up being a dumping ground for anything.

So I suggest we add a placeholder variable that can be set in `env.local` to dynamically add extra xml snippets to add any number of new catalogs.  This will cover Deepak 3 catalogs suggestions and any other special needs for any future node admin.  Nothing says 3 catalogs is enough or 5 or 10.  Each node admin will decide how many catalogs they want.  We start with 2 just to show that it is possible to have more than 1 catalog.

Example of placeholder type of variable: https://github.com/bird-house/birdhouse-deploy/blob/33042e6b0b1cc46ca0ec18bffca9d74b4d385f53/birdhouse/config/jupyterhub/jupyterhub_config.py.template#L190 

Sample use of that placeholder var in `env.local`: https://github.com/bird-house/birdhouse-deploy/blob/33042e6b0b1cc46ca0ec18bffca9d74b4d385f53/birdhouse/env.local.example#L338-L350

Reply:
Yes, there is absolutely no need to constrain ourselves to a pre-determined number of catalogs. This sounds like a very good way to proceed. For now, on PAVICS, the `RAW_DATA` and `DATA` catalogs would map directly to the existing PAVICS setup and for UofT the `RAW_DATA` would contain, among other things, (i) temporary data that we don't want to host long-term and so we don't want to put in the main catalog, (ii) data that needs to be quickly made available to users for their analysis while we work on properly ingesting it into the main catalog.

Reply:
To not constraint any data type in any catalog, how about renaming from `RAW_DATA` and `DATA` flavors to simply `CATALOG_1` and `CATALOG_2`.   Given the generic variable name, the default values will look funny.  Just encourage users to set different names for them in the documentation.

Reply:
@tlvu 
>   Just document they are just 2 catalogs, nothing special, the admin will decide what goes in there. It could be just testdata, just wps_outputs, not a mixed bag of everything like "Birdhouse" is right now.

I strongly disagree with this.

The plateform itself, especially with Cowbird, tries to establish "shared" locations between services, and synchronize their permissions. 
It is already hard enough to make everything work with the various interfaces of all these services, I would like at least to have 1 directory that is "platform-managed". So maybe `PLATFORM_DATA` or `SERVICE_DATA` could work as well instead of `BIRDHOUSE_DATA`. Basically, I would like the name to make it stand out for admins as "*you should not host your data here, services could wipe it*", and avoid this mixed uses that were listed. It over-complicates how Cowbird needs to work around different locations and filesystem structures without breaking permissions. 

For the other data locations, whether there is 1, 2 or more, and whether it is called "datasets" or anything else doesn't matter that much for me, because it is for data-dumping whatever the node admin wants to share.

> So I suggest we add a placeholder variable that can be set in env.local to dynamically add extra xml snippets to add any number of new catalogs. 

That would be a good way to help node admins add more dataset locations as needed without completely overriding the current catalog XML.

Reply:
Ok, I'm going to try to summarize where I think we're currently at with this discussion:

1. We need one location for data that is created/managed by components like cowbird and weaver  
    - Let's use `SERVICE` in the variable names for this location (I don't have a strong preference but I think we just need to pick one)
2. We can have any other number of locations that can be added by adding extra xml snippets
     - note that we will also need a mechanism to specify which locations on the host should be mounted to the thredds container but we can probably do that by setting variables in env.local as well

Any objections to this strategy?

Reply:
Agreed with 1 of the 2 location called `SERVICE` to denote its shared service nature.  Mention that in the documentation.

Reply:
Sounds good. We can have 1 `SERVICE` and 1 `DATASET` to make 2 locations quickly avaiable (without need of custom/complicated XML override), and N extentable locations to the admin's discretion using a 3rd variable.

Reply:
If we are deciding that one of the two locations will be called `SERVICE` then we will need the ability to have two datasets right from the start. So, I will have to include the XML extension ability in this PR as well. Give me some time to figure that out.

Reply:
@dchandan The ability to have extra catalogs is simply a variable in `env.local` that will be template expanded into the Thredds config file.  Should be really simple, unless there is template expansion problem.

For the 2 existing catalogs, hardcode them to make things simple.  Do not make the 2nd catalog configurable. 

A new org deploying PAVICS, will simply put all its data into the second `DATASET` catalog and not touch the `SERVICE` catalog.  In the `SERVICE` catalog, the only item is `wps_outputs` that are shared by all the birds and Cowbird.

Reply:
>  the only item is `wps_outputs`

And `testdata` as well IMO, but I agree with @tlvu for the rest.

Reply:
> unless there is template expansion problem.

Yeah, that is where I expect any potential problems to arise. Hopefully, not.

Reply:
If the variable uses `'` quotes, there shouldn't be issues to inject XML as is, since it would be dropped in without shell expansions, but that would not allow for it to nest other template variables easily. Otherwise, `"` are possible but they must deal with any necessary escapes since `"` are used extensively in XML. 

Alternatively, including another XML file could be possible if THREDDS's XML parser allows it (not sure if it does?). 
```xml
<!DOCTYPE doc [
<!ENTITY otherFile SYSTEM "otherFile.xml">
]>
&otherFile;
```
Assuming this is allowed, the admin could then volume mount that XML, which could be generated from a corresponding `otherFile.xml.template`, and not have to worry about any escapes.

Reply:
Done. I hope [this implementation](https://github.com/bird-house/birdhouse-deploy/pull/359/commits/8c5f78c27dfba47c18ba7b34918cb1df2e013923) makes sense.